### PR TITLE
Deep clone TaskRegistry and Toolset

### PR DESCRIPTION
**Pull request by JanKrivanek**
Fixes [ADO#1801351](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1801351) and [ADO#1801341](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1801341) and possibly [ADO#1824802](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1824802)

Credit for rootcausing: @rokonec 

### Context
`TaskRegistry` (and as `Toolset` as well) were not properly clonned during clonning `ProjectInstance` data - that could cause unintended sharing of `TaskRegistry` data between multiple independent evaluation/build threads - allowing concurrent access to `TaskRegistry` and its objects, while such concurency is not supported.

### Changes Made
Made `TaskRegistry` and `Toolset` classes DeepClonable and properly cloning them during `ProjectInstance` clonning

### Testing - Perf

No observable impact.

 * OrchardCore and Console, build command: `msbuild.exe /nodeReuse:false /bl`

| Scenario | Mean Duration |
| --- | --- |
| Orchard, MSBuild main, clean build, /bl | 00:01:54.28 |
| Orchard, MSBuild curr branch, clean build, /bl | 00:02:01.36 |
|  |  |
| Orchard, MSBuild main, rebuild, /bl | 00:01:44.46 |
| Orchard, MSBuild curr branch, rebuild, /bl | 00:01:41.81 |
|  |  |
| Orchard, MSBuild main, rebuild | 00:01:36.55 |
| Orchard, MSBuild curr branch, rebuild | 00:01:39.83 |
| --- | --- |
| Console, MSBuild main, clean build | 00:00:01.75 |
| Console, MSBuild curr branch, clean build | 00:00:01.71 |
|  |  |
| Console, MSBuild main, rebuild | 00:00:01.34 |
| Console, MSBuild curr branch, rebuild | 00:00:01.31 |

### CODE REVIEW

File: src/Build/Definition/Toolset.cs [hunk 1]
```diff
@@ -901,79 +920,78 @@ private void RegisterDefaultTasks(ILoggingService loggingServices, BuildEventCon
...
+                int count = _environmentProperties.Count + reservedProperties.Count + Properties.Values.Count + _globalProperties.Count + 1;
 
-                    // GenerateSubToolsetVersion checks the environment and global properties, so it's safe to go ahead and gather the
-                    // subtoolset properties here without fearing that we'll have somehow come up with the wrong subtoolset version.
-                    string subToolsetVersion = this.GenerateSubToolsetVersion();
-                    SubToolset subToolset;
-                    ICollection<ProjectPropertyInstance> subToolsetProperties = null;
+                // GenerateSubToolsetVersion checks the environment and global properties, so it's safe to go ahead and gather the
+                // subtoolset properties here without fearing that we'll have somehow come up with the wrong subtoolset version.
+                string subToolsetVersion = this.GenerateSubToolsetVersion();
```
Comment (donJoseLuis) [inline src/Build/Definition/Toolset.cs:956]: minor: is the `this.*` needed?

**Review by donJoseLuis — APPROVED** [review]

File: src/Build/Definition/Toolset.cs [hunk 2]
```diff
@@ -190,11 +189,6 @@ public class Toolset : ITranslatable
         /// </summary>
         private Expander<ProjectPropertyInstance, ProjectItemInstance> _expander;
 
-        /// <summary>
-        /// Bag of properties for the expander to expand the properties and items in the using tasks files
-        /// </summary>
-        private PropertyDictionary<ProjectPropertyInstance> _propertyBag;
```
Comment (rokonec) [inline src/Build/Definition/Toolset.cs:196]: ❤️

File: src/Build/Definition/Toolset.cs [hunk 3]
```diff
@@ -901,79 +920,78 @@ private void RegisterDefaultTasks(ILoggingService loggingServices, BuildEventCon
         /// </summary>
         private void InitializeProperties(ILoggingService loggingServices, BuildEventContext buildEventContext)
         {
+            if (_expander != null)
```
Comment (rokonec) [inline src/Build/Definition/Toolset.cs:923]: ❤️

File: src/Build/Instance/ProjectInstance.cs [hunk 4]
```diff
@@ -561,9 +561,13 @@ internal ProjectInstance(Evaluation.Project.Data data, string directory, string
             this.CreateTargetsSnapshot(data.Targets, data.DefaultTargets, data.InitialTargets, data.BeforeTargets, data.AfterTargets);
             this.CreateImportsSnapshot(data.ImportClosure, data.ImportClosureWithDuplicates);
 
-            this.Toolset = data.Toolset; // UNDONE: This isn't immutable, should be cloned or made immutable; it currently has a pointer to project collection
             this.SubToolsetVersion = data.SubToolsetVersion;
-            this.TaskRegistry = data.TaskRegistry;
+            this.TaskRegistry = data.TaskRegistry?.DeepClone();
+
+            // If the task registry uses toolset identical to the one in data instance - deep clone it just once.
```
Comment (rokonec) [inline src/Build/Instance/ProjectInstance.cs:567]: Please explain why - in comment.  The fact that it is about reusing already cloned instance by line above might  not be obvious for reader.

File: src/Build/Instance/ProjectInstance.cs [hunk 5]
```diff
@@ -641,11 +645,12 @@ private ProjectInstance(ProjectInstance that, bool isImmutable, RequestedProject
                     ProjectItemDefinitionInstance>)this).AfterTargets = CreateCloneDictionary(
                     ((IEvaluatorData<ProjectPropertyInstance, ProjectItemInstance, ProjectMetadataInstance,
                         ProjectItemDefinitionInstance>)that).AfterTargets, StringComparer.OrdinalIgnoreCase);
-                this.TaskRegistry =
-                    that.TaskRegistry; // UNDONE: This isn't immutable, should be cloned or made immutable; it currently has a pointer to project collection
 
-                // These are immutable so we don't need to clone them:
-                this.Toolset = that.Toolset;
+                this.TaskRegistry = that.TaskRegistry.DeepClone();
+                // If the task registry uses toolset identical to the one in project instance - deep clone it just once.
```
Comment (rokonec) [inline src/Build/Instance/ProjectInstance.cs:650]: Same as above please update comment to state that we are reusing already cloned Toolset

**Review by rokonec — APPROVED** [review]
Good job.
This would be better in another PR, but I believe that we shall alter our tests so it would catch common failure mode when people add field/property and forget to Clone it. 
We have unit tests comparer but those will most probably be forgotten to update as well.

I had success in pasty with [AutoFixture](https://adrianhesketh.com/2014/10/08/unit-testing-mapping-and-serialization/) for seeding data and [Fluent Assertion equivalency tests](https://fluentassertions.com/objectgraphs/)

File: src/Build/Definition/Toolset.cs (hunk 1 above)
Comment (JanKrivanek) [reply src/Build/Definition/Toolset.cs:956]: Removed

File: src/Build/Definition/ProjectImportPathMatch.cs [hunk 6]
```diff
@@ -24,36 +32,39 @@ internal ProjectImportPathMatch(string propertyName, List<string> searchPaths)
...
+            _searchPaths = searchPaths;
+            _msBuildPropertyFormat = $"$({PropertyName})";
         }
 
         public ProjectImportPathMatch(ITranslator translator)
         {
             ((ITranslatable)this).Translate(translator);
         }
 
+        internal ProjectImportPathMatch DeepClone()
```
Comment (rainersigwald) [inline src/Build/Definition/ProjectImportPathMatch.cs:45]: If it's immutable, does it need to be cloned?

Comment (JanKrivanek) [reply src/Build/Definition/ProjectImportPathMatch.cs:45]: ah good point - removing...

Comment (JanKrivanek) [reply src/Build/Definition/ProjectImportPathMatch.cs]: (removed)

File: src/Build.UnitTests/Instance/ProjectInstance_Internal_Tests.cs [hunk 7]
```diff
@@ -480,8 +481,8 @@ public void CloneTaskRegistry()
             ProjectInstance first = GetSampleProjectInstance();
             ProjectInstance second = first.DeepCopy();
 
-            // Task registry object should be immutable
```
Comment (rainersigwald) [inline src/Build.UnitTests/Instance/ProjectInstance_Internal_Tests.cs:483]: Was this assertion always wrong? Have we added mutability where we shouldn't have, or just "in a way that requires we change this"?

Comment (JanKrivanek) [reply src/Build.UnitTests/Instance/ProjectInstance_Internal_Tests.cs:483]: Unfortunately this is older then GH history so I can only provide guessing - and my guess is the ProjectInstance and owned object graph (including TaskRegistry) was always meant to be instantiated once and then used only sequantially. So immutability and deepclonning was never strongly followed - as it was mostly ok in majority of cases (a specific set of VS actions and conditions needs to be made - e.g. in-proc build or cancel and rebuild - to hit the unwanted sharing of TaskRegistry/Toolset).
TaskRegistry looks _almost_ immutable, with exception of the internal matching caches, lazy initialized state of held TaskFactoryWrapper instances (the reason for the 3rd state corruption perf issue), plus something else (I forgot now) - so the path of DeepCopying turned out to be easier and safer than attempting to make TaskRegistry and Toolset truly immutable

Comment (JanKrivanek) [reply src/Build.UnitTests/Instance/ProjectInstance_Internal_Tests.cs:483]: tl;dr; the added mutability shouldn't be added without adding deep cloning IMHO

**Review by rainersigwald — COMMENTED** [review]
Looking at this more, this feels wrong to me. I expect the Toolset and TaskRegistry to be essentially singletons (since we no longer respect a file's toolset specification), and now there will be many of them, causing us to have a lower cache hit rate on lookups.

Do we understand yet what's causing the multithreading in task execution? Or is it one thread executing tasks and one thread evaluating projects? Or what?

**Comment by JanKrivanek** [comment]
> Looking at this more, this feels wrong to me. I expect the Toolset and TaskRegistry to be essentially singletons (since we no longer respect a file's toolset specification), and now there will be many of them, causing us to have a lower cache hit rate on lookups.
> 
> Do we understand yet what's causing the multithreading in task execution? Or is it one thread executing tasks and one thread evaluating projects? Or what?

Copying ProjectInstances is exposed publicly (with coment mentioning it's deep copy) - so there might be scenarios we don't know about. Overlaping build and evaluation seems to be one of them (as copy ctor is invoked between eval and build). The mutability concern and need for deep copy was voiced in the original code as well: https://github.com/dotnet/msbuild/pull/8973/files#diff-40f77147ccc6867c10324ec4ddf8110ede2e43499d939b73ef83c7f5febb7a75L564

However, let's make sure we have better understanding and better consensus on solution - so keeping this PR open for now.

File: src/Build/Definition/SubToolset.cs [hunk 8]
```diff
@@ -50,6 +50,9 @@ private SubToolset(ITranslator translator)
             ((ITranslatable)this).Translate(translator);
         }
 
+        internal SubToolset DeepClone()
+            => new SubToolset(_subToolsetVersion, _properties?.DeepClone());
```
[the remark this answers is not in the document]
Comment (JanKrivanek) [reply src/Build/Definition/SubToolset.cs:54]: :thumbsup:
The code could (and likely should) verify the `IsImmutable` member of all `_properties` and skip clonning if they are all `true`.
Without that unfortunately the code gives no guarantees.

Or possibly the SubToolset itself should be created immutable (with properties flipped to immutable) - as currently this is not a case

**Comment by JanKrivanek** [comment]
Upon detailed investigation with @rainersigwald and @rokonec we'll fix this differently: TaskRegistry is expected to be shareable and logicaly immutable, however it has internal caches, that can be altered on access. so TaskRegistry internal caches will be made thread safe, the actual task registrations are not expected to be changed after initialization - so safeguard code will be added to catch unintended state changes

**Comment by JanKrivanek** [comment]
Superseded by https://github.com/dotnet/msbuild/pull/9032

**Pull request closed without merging**